### PR TITLE
applications: serial_lte_modem: Add #XSLEEP=2 (power off UART)

### PR DESCRIPTION
**Pull request by junqingzou**
Customer requires powering off UART only, plus wake-up from MCU.
Change #XSLEEP=1 behavior so that CONFIG_SLM_GPIO_WAKEUP is required.
Change default configuration as below:                                                                                  
.CONFIG_SLM_GPIO_WAKEUP is defined
.new CONFIG_SLM_START_SLEEP is not defined
To sum up, support deep sleep/wakeup while SLM starts up as active.
Also removed the dependency on LTE Link Control library.

JIRA Reference: NCSIDB-386

Signed-off-by: Jun Qing Zou <jun.qing.zou@nordicsemi.no>

### CODE REVIEW

**Comment by rlubos** [comment]
@anangl Could you have a look, I don't feel very confident in UART/PM area.

File: applications/serial_lte_modem/src/slm_at_host.c [hunk 1]
```diff
@@ -298,22 +326,33 @@ static int handle_at_sleep(const char *at_cmd, enum shutdown_modes *mode)
...
 		if (shutdown_mode == SHUTDOWN_MODE_IDLE) {
 			slm_at_host_uninit();
-			enter_idle();
+			enter_idle(true);
 			*mode = SHUTDOWN_MODE_IDLE;
 			ret = 0; /*Will send no "OK"*/
 		} else if (shutdown_mode == SHUTDOWN_MODE_SLEEP) {
 			slm_at_host_uninit();
-			enter_sleep(true);
+			enter_sleep(false);
```
Comment (anangl) [inline applications/serial_lte_modem/src/slm_at_host.c:334]: Was that change intended? The commit message does not mention anything about changing the behavior of `AT#XSLEEP=1`.

File: applications/serial_lte_modem/src/slm_at_host.c [hunk 2]
```diff
@@ -298,22 +326,33 @@ static int handle_at_sleep(const char *at_cmd, enum shutdown_modes *mode)
...
 			ret = 0; /* Cannot reach here */
```
Comment (anangl) [inline applications/serial_lte_modem/src/slm_at_host.c:335]: This comment seems to be outdated. Even `enter_sleep(true)` can return when `CONFIG_SLM_GPIO_WAKEUP` is not enabled.

File: applications/serial_lte_modem/src/slm_at_host.c (hunk 1 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_host.c:334]: Thanks a lot for this, don't know why I messed up here.

File: applications/serial_lte_modem/src/slm_at_host.c (hunk 2 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/slm_at_host.c:335]: Yes, updated

**Comment by junqingzou** [comment]
DNM due to this PR is to be tested by customer first, and possible more feature request.

**Review by anangl — APPROVED** [review]
LGTM

**Review by lats1980 — COMMENTED** [review]
How about also wakeup uart if there is URC to be transferred UART? Then MQTT subsrciption/TCP proxy received message can be notified to external MCU:

void rsp_send(const uint8_t *str, size_t len)
{
	int ret;

	if (len == 0) {
		return;
	}

	if(wakeup_uart() != 0) {
		LOG_WRN("Fail to wake up uart");
		return;
	}
	k_sem_take(&tx_done, K_FOREVER);

	uart_tx_buf = k_malloc(len);
	if (uart_tx_buf == NULL) {
		LOG_WRN("No ram buffer");
		k_sem_give(&tx_done);
		return;
	}

	LOG_HEXDUMP_DBG(str, len, "TX");

	memcpy(uart_tx_buf, str, len);
	ret = uart_tx(uart_dev, uart_tx_buf, len, SYS_FOREVER_MS);
	if (ret) {
		LOG_WRN("uart_tx failed: %d", ret);
		k_free(uart_tx_buf);
		k_sem_give(&tx_done);
	}
}

**Comment by junqingzou** [comment]
> How about also wakeup uart if there is URC to be transferred UART? Then MQTT subsrciption/TCP proxy received message can be notified to external MCU:
> 
> void rsp_send(const uint8_t *str, size_t len)
> {
> int ret;
> 
> ```
> if (len == 0) {
> 	return;
> }
> 
> if(wakeup_uart() != 0) {
> 	LOG_WRN("Fail to wake up uart");
> 	return;
> }
> k_sem_take(&tx_done, K_FOREVER);
> 
> uart_tx_buf = k_malloc(len);
> if (uart_tx_buf == NULL) {
> 	LOG_WRN("No ram buffer");
> 	k_sem_give(&tx_done);
> 	return;
> }
> 
> LOG_HEXDUMP_DBG(str, len, "TX");
> 
> memcpy(uart_tx_buf, str, len);
> ret = uart_tx(uart_dev, uart_tx_buf, len, SYS_FOREVER_MS);
> if (ret) {
> 	LOG_WRN("uart_tx failed: %d", ret);
> 	k_free(uart_tx_buf);
> 	k_sem_give(&tx_done);
> }
> ```
> 
> }

rsp_send() is called by many many URCs so should not add this step. Actually there is no intention of AUTOMATIC wakeup by URCs from SLM. The wakeup request from SLM inside should be explicitly made only when needed (i.e. being sure that UART has been powered off).

**Review by rlubos — APPROVED** [review]

**Review by lats1980 — APPROVED** [review]
Approve!

**Review by lats1980 — APPROVED** [review]
Looks good to me.

**Comment by junqingzou** [comment]
@rlubos @anangl @lats1980  Please review again as there is some more changes for customer project

**Review by anangl — APPROVED** [review]

**Comment by ru-fu** [comment]
Does step 4 here need any updates?
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/serial_lte_modem/doc/slm_testing.html#generic-at-commands

**Comment by ru-fu** [comment]
Should LTE link controller be removed from the dependencies?
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/serial_lte_modem/doc/slm_description.html#dependencies

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst [hunk 3]
```diff
@@ -135,6 +135,7 @@ The ``<shutdown_mode>`` parameter accepts only the following integer values:
 * ``0`` - Enter Idle.
   You can also use the syntax ``AT#XSLEEP``.
 * ``1`` - Enter Sleep.
+* ``2`` - Power off UART.
 
 The default value is 0.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/Generic_AT_commands.rst:140]: Need to add the behavior for power off to the list in line 146/147,

File: applications/serial_lte_modem/doc/slm_description.rst [hunk 4]
```diff
@@ -135,10 +135,17 @@ Check and configure the following configuration options for the sample:
 
 .. option:: CONFIG_SLM_GPIO_WAKEUP - Support of GPIO wakeup
 
-   This option enables using GPIO to wake up from sleep mode.
-   Select this option if you want to test the application with an external CPU.
+   This option enables using GPIO to wake up nRF91 from deep sleep mode.
+   Select this option if you want to support deep sleep of nRF91.
+
+   Note this option is by default selected.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/slm_description.rst:141]:
```suggestion
   This option enables using GPIO to wake up nRF9160 from deep sleep mode.
   Select this option if you enable :option:`CONFIG_SLM_START_SLEEP` to put nRF9160 into deep sleep mode after startup.

   This option is selected by default.
```

File: applications/serial_lte_modem/doc/slm_description.rst [hunk 5]
```diff
@@ -135,10 +135,17 @@ Check and configure the following configuration options for the sample:
...
+
+.. option:: CONFIG_SLM_START_SLEEP - Enter sleep on start-up
+
+   This option makes nRF91 enter deep sleep after power on.
+   This option is by default not selected.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/slm_description.rst:146]:
```suggestion
.. option:: CONFIG_SLM_START_SLEEP - Enter sleep on startup

   This option makes nRF9160 enter deep sleep after startup.
```

File: applications/serial_lte_modem/doc/slm_description.rst [hunk 6]
```diff
@@ -135,10 +135,17 @@ Check and configure the following configuration options for the sample:
...
 
-   If this option is not selected, you must reset the kit to exit sleep mode.
+   Note this option requires :option:`CONFIG_SLM_GPIO_WAKEUP` is selected.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/slm_description.rst:148]:
```suggestion
   This option is not selected by default.
   It requires :option:`CONFIG_SLM_GPIO_WAKEUP` to be selected.
```

File: applications/serial_lte_modem/Kconfig [hunk 7]
```diff
@@ -80,9 +80,19 @@ config SLM_AT_HOST_TERMINATION
...
+	default y
 	help
 	  Enable GPIO wakeup on nRF9160 side.
 
+if SLM_GPIO_WAKEUP
+
+config SLM_START_SLEEP
+	bool "Enter sleep on start-up"
+	help
+	  nRF91 start up then enter deep sleep.
```
Comment (ru-fu) [inline applications/serial_lte_modem/Kconfig:92]:
```suggestion
	bool "Enter sleep on startup"
	help
	  Put nRF9160 into deep sleep mode after startup.
```

**Review by ru-fu — CHANGES_REQUESTED** [review]

File: applications/serial_lte_modem/doc/slm_description.rst (hunk 4 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/slm_description.rst:141]: Updated

File: applications/serial_lte_modem/doc/slm_description.rst (hunk 5 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/slm_description.rst:146]: Updated

File: applications/serial_lte_modem/doc/slm_description.rst (hunk 6 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/slm_description.rst:148]: Updated

File: applications/serial_lte_modem/Kconfig (hunk 7 above)
Comment (junqingzou) [reply applications/serial_lte_modem/Kconfig:92]: Updated

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst (hunk 3 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/Generic_AT_commands.rst:140]: Updated

**Review by rlubos — APPROVED** [review]

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst [hunk 8]
```diff
@@ -143,8 +144,9 @@ Response syntax
 
 There is no response:
 
-* In case of Idle, it will exit by sending data over UART.
-* In case of Sleep, it will wake up by GPIO or reset.
+* In case of Idle, it will exit by GPIO.
+* In case of Sleep, it will wake up by GPIO.
+* in case of UART power off, it will be powered on by GPIO or by SLM when needed.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/Generic_AT_commands.rst:149]:
```suggestion
* In case of UART power off, it will be powered on by GPIO or by SLM when needed.
```

**Review by ru-fu — COMMENTED** [review]
> Does step 4 here need any updates?
> https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/applications/serial_lte_modem/doc/slm_testing.html#generic-at-commands
> 
> Yes updated

I can't see an update in that step, but I don't know if it's needed.

One more small fix.

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst [hunk 9]
```diff
@@ -132,9 +132,10 @@ Syntax
 
 The ``<shutdown_mode>`` parameter accepts only the following integer values:
 
-* ``0`` - Enter Idle.
+* ``0`` - Enter Idle, SLM service terminated but LTE connection maintained.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/Generic_AT_commands.rst:135]:
```suggestion
* ``0`` - Enter Idle.
  In this mode, the SLM service is terminated, but the LTE connection is maintained.
```

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst [hunk 10]
```diff
@@ -132,9 +132,10 @@ Syntax
...
   You can also use the syntax ``AT#XSLEEP``.
-* ``1`` - Enter Sleep.
+* ``1`` - Enter Sleep, both SLM service and LTE connection terminated.
+* ``2`` - Power off UART, both SLM service and LTE connection maintained.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/Generic_AT_commands.rst:138]:
```suggestion
* ``1`` - Enter Sleep.
  In this mode, both the SLM service and the LTE connection are terminated.
* ``2`` - Power off UART.
  In this mode, both the SLM service and the LTE connection are maintained.
```

File: applications/serial_lte_modem/doc/slm_testing.rst [hunk 11]
```diff
@@ -61,15 +61,20 @@ Complete the following steps to test the functionality provided by the :ref:`SLM
       :class: highlight
 
       **AT#XSLEEP=?**
-      #XSLEEP: (0,1)
+      #XSLEEP: (0,1,2)
       OK
 
+      **AT#XSLEEP**
+
+   You can quit the Idle status by GPIO.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/slm_testing.rst:69]:
```suggestion
   ``AT#XSLEEP`` puts the kit in idle mode, and you can wake up by GPIO.
   
   Alternatively, you can use different modes for #XSLEEP:
   
   * ``AT#XSLEEP=1`` puts the kit in sleep mode.
     If you are testing with :option:`CONFIG_SLM_GPIO_WAKEUP` enabled, you can wake up by GPIO.
   * ``AT#XSLEEP=2`` powers off UART.
     You can power on UART by GPIO.
```

File: applications/serial_lte_modem/doc/slm_testing.rst [hunk 12]
```diff
@@ -61,15 +61,20 @@ Complete the following steps to test the functionality provided by the :ref:`SLM
...
+
       **AT#XSLEEP=1**
 
-   Reset the kit to exit sleep mode.
-   If you are testing with an external MCU and :option:`CONFIG_SLM_GPIO_WAKEUP` is enabled, you can wake up by GPIO as well.
+   If you are testing with :option:`CONFIG_SLM_GPIO_WAKEUP` enabled, you can wake up by GPIO.
 
+      **AT#XSLEEP=2**
 
+   You can power on UART by GPIO.
```
Comment (ru-fu) [inline applications/serial_lte_modem/doc/slm_testing.rst:77]: Remove these (covered in the suggestion above).

Comment (junqingzou) [reply applications/serial_lte_modem/doc/slm_testing.rst]: Updated

File: applications/serial_lte_modem/doc/slm_testing.rst (hunk 11 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/slm_testing.rst]: Updated

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst (hunk 10 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/Generic_AT_commands.rst]: Updated

File: applications/serial_lte_modem/doc/Generic_AT_commands.rst (hunk 9 above)
Comment (junqingzou) [reply applications/serial_lte_modem/doc/Generic_AT_commands.rst]: Updated

**Review by ru-fu — APPROVED** [review]
Docs look good!
